### PR TITLE
[nikon] Add Tokina AT-X 14-20 F2 PRO DX lens

### PR DESCRIPTION
**Pull request by Sturmflut**
Came in via lensfun/lensfun#1188

### CODE REVIEW

**Review by clanmills — COMMENTED** [review]
That test code is ugly.  Your cmd array has one command.  The stdout uses python `"" """..."""` which I believe python concatenates into one string.  However it's ugly.   No idea why that would be different on Windows.

File: tests/bugfixes/github/test_pr_2073.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+
+import system_tests
+
+class NikonSigmaLens_APO_MACRO_180_F35_EX_DG_HSM(metaclass=system_tests.CaseMeta):
+    url = "https://github.com/Exiv2/exiv2/pull/992"
+    
+    filename = "$data_path/Sigma_APO_MACRO_180_F3.5_EX_DG_HSM.exv"
```
Comment (piponazo) [inline tests/bugfixes/github/test_pr_2073.py:8]: There seems to be a copy/paste typo here. You are using a different file than the one added in this PR

**Review by piponazo — CHANGES_REQUESTED** [review]

**Comment by Sturmflut** [comment]
@clanmills I was following https://github.com/Exiv2/exiv2/wiki/Adding-a-new-lens , but it isn't very helpful in this regard, simply pointing to other issues and discussions. I ended up using the test from #992 because it looked as good as anything else.

It looks like the whole `"" """...""" ""` construction is just there to emulate a carriage return, which can also simply be done using `\n`.

While going over this one and the mistakes I had made in #2072 and #2069, I realised that

1. The `bugfix` section is not the right one for these simple tests, there is already a lens_tests section
2. The tests for Nikon lenses should be consolidated into a single file, like `test_canon_lenses.py`

The Canon tests work by extracting the data from `canonmn_int.cpp`, then feeding the fields to an exiv2 command by artificially setting the right fields, and looking if the lens is detected correctly.

That works and catches bugs, but the data is only validated against itself. Accidental changes to the data structure will not result in a failed test.

I propose a new file called `test_nikon_lenses.py` in `tests/lens_tests` that creates tests from an array of known EXIF files in `test/data/` and the expected lens names. I've updated the branch with an example for the lens in this pull request.

**Comment by clanmills** [comment]
@Sturmflut Thanks for this.

Before I get into specifics, please allow me to say "Welcome to Exiv2".  I'm very pleased to see you here and making useful contributions.

I did approve #992.  I should have requested that the ugly code be either:

```python
    stdout = ["""Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma APO MACRO 180mm F3.5 EX DG
 """]
 ```
or
```python
    stdout = ["""Exif.NikonLd3.LensIDNumber                   Byte        1  Sigma APO MACRO 180mm F3.5 EX DG \n"""]
```

The C language (and therefore C++) has always had the string concat syntax:
```
  string := "...."
         | string string
  ;
```
python has the additional multi-line string `"""..."""`

By co-incidence I was discussing  https://github.com/Exiv2/exiv2/wiki/Adding-a-new-lens with @postscript-dev yesterday.  For sure, that document needs effort.  I wrote it in frustration with "the lens problem".  We get about one bug report every week about lens recognition.  Lens recognition isn't metadata, it involves guessing the lens from other metadata.  The Exif Specification fairly recently added Exif.Photos.LensModel and Exiv2 support this tag.  My agenda in writing the document was to make it clear that the request to  _"update your database with this lens"_  isn't trivial.

I added the "configuration file" in v0.26.  That's the quick and easy way to add/modify lens recognition.

If you'd like to edit/revise https://github.com/Exiv2/exiv2/wiki/Adding-a-new-lens, both Peter (@postscript-dev) and I welcome your contribution and will review your changes.

I was unaware of `tests/lens_tests` which has been added to branch `main`.  I maintain 0.27-maintenance which does not have such a directory.  Your proposal to add nikon related lenses seems good to me.

I started to revise tests/README-TESTS.md and couldn't complete that due to time-pressure.  https://github.com/Exiv2/exiv2/pull/1553

I see CodeCov has correctly (and ingeniously) detected changes in code coverage.  That should be an amber (warning) and not a red (failing) flag.  

If you'd like to revise https://github.com/Exiv2/exiv2/wiki/Adding-a-new-lens, I'll be happy to grant you the GitHub permissions to modify that document.  If you prefer not to get further involved with Exiv2, I thank you for your contributions and hope you'll join us in the future.

**Comment by clanmills** [comment]
About the CodeCov "failures".  I think CodeCov has spotted something very useful.  Your changes have made some code obsolete and should probably be removed from the code base.  Can you consider if some of the code in `src/nikonmn_int.cpp` can be removed?

**Comment by hassec** [comment]
I think the discussion is maybe getting a bit sidetracked here.   

Let's turn this PR back into a simple "Add Tokina AT-X 14-20 F2 PRO DX lens" one, similar to what you did in e.g. #2069. 
(maybe also try and see if the suggestion from #2072 works and use that here as well)  

Then re: 
>The Canon tests work by extracting the data from canonmn_int.cpp, then feeding the fields to an exiv2 command by artificially setting the right fields, and looking if the lens is detected correctly.                  
That works and catches bugs, but the data is only validated against itself. Accidental changes to the data structure will not result in a failed test.

That's correct but those are **only part** of the canon lens testing we do.     
We also have the same tests for canon you add here for nikon, that is small `.exv` files that we decode and check that we get the correct lens name as result. These tests are important, and would be all we need if we had a complete set of files for every single lens.  

But because we don't have that we additionally use the the above test to make sure that we can do the roundtrip of "Lens  specification -> exif values -> lens name" which isn't always guaranted because we deduce the lens name based on certain values like aperture etc, which are encoded in the exif data by a lossy transformation and thus make the roundtrip lens recognition sometimes a bit challenging. But essentially it's:     
"if Canon encodes data in a way we expect, can we determine the lens"                  
(That's the short version, happy to give you the longer one if you are interested :joy: )   

So for Nikon we really should implement the same test to, but its a bit of work and after having implemented the test for canon I've not yet had the time to do that.     

All the above aside, the suggestion you made of having a single file that has a map of "Filename -> Lens Name" 
and thus defines a single data driven lens recogintion test per maker, is IMHO a great idea! :+1:     
But I'd suggest we do that in a new PR where we go through the exisiting tests we have that e.g. check for Nikon lenses and merge all of them into one test in the format you suggest here.   

And then we can do the same for other brands but that can also be done step by step :wink:

**Comment by clanmills** [comment]
@hassec You're right, Christoph.   @Sturmflut is doing a good job.   I've wandered off-piste and will keep out of this issue.

**Comment by piponazo** [comment]
> I propose a new file called `test_nikon_lenses.py` in `tests/lens_tests` that creates tests from an array of known EXIF files in `test/data/` and the expected lens names. I've updated the branch with an example for the lens in this pull request.

@Sturmflut I really like this new approach for testing the nikon lenses :+1: 

What do you prefer to do for the 2 existing PRs? Should we merge this one and then you adapt #2072 , or do you prefer to include the changes from #2072 here?

**Review by piponazo — DISMISSED** [review]

**Comment by Sturmflut** [comment]
I changed this back to use the same simple test as in #2072. I'll send a new pull request with a proposal for the tests.

**Comment by Sturmflut** [comment]
@clanmills I'll happily update the wiki page, regardless of what we decide in #2071

**Review by piponazo — APPROVED** [review]

**Comment by clanmills** [comment]
Thanks, @Sturmflut.  I happy to keep out of this.  You're doing fine.

**Pull request merged**